### DISCUSSION TITLE
VSCode Extension for IF

### DISCUSSION THREAD

**Discussion by andrew-woosnam** [Green-Software-Foundation/if#737] [Show and tell]
Hey friends!

I've been collaborating with @josh-swerdlow on [this Visual Studio Code extension](https://github.com/andrew-woosnam/impact-framework-vscode/tree/main) that aims to make working with Impact Framework easier. It's very much a work in progress, but we'd love to hear from you all in terms of whether or not this endeavor is worth pursuing further.

You can learn more in the [readme](https://github.com/andrew-woosnam/impact-framework-vscode/blob/main/README.md), but so far we've added these capabilities:
- Create new IF manifest file via vscode command.
- Snippets to autocomplete IF manifest file sections.
- Hover-over descriptions of IF manifest file components.
- Sidebar listing available IF plugins.

## Request for Feedback
We would greatly appreciate your feedback on the current state of this MVP. Specifically, we are looking for insights on:

- [ ] **Usefulness**: How useful are the existing features in your workflow? Are there any immediate improvements you would suggest?

- [ ] **Additional Features**: What other features would you like to see in this extension? Please share any ideas or pain points that we can address in future updates.

> Note: We are aware of several areas for improvement, such as:
> - Shifting from hardcoded text to dynamically queried content.
> - Enhancing plugin management, including the ability to handle plugin installations (e.g., via npm install) directly from the plugins sidebar.

**Comment by andrew-woosnam** [comment]
@heaversm, pinging you here since you seemed interested in a vscode extension in an [earlier conversation](https://github.com/Green-Software-Foundation/hack/discussions/136#discussioncomment-9335411) with @josh-swerdlow

**Reply by heaversm** [reply]
Thanks - I'm not seeing this in the vscode registry, or a vsix file, so I'm assuming to test this out we'd need something like vsce and package this ourselves and then do a manual install?

**Reply by andrew-woosnam** [reply]
Ah apologies, yeah it's not on any marketplace and I haven't cut any "release" yet (so I really should've at least provided instructions on how to build it yourself 🙄 my bad!)

I'm updating the repo now with a .vsix + some build instructions, stay tuned ...

**Reply by andrew-woosnam** [reply]
Here you go @heaversm: https://github.com/andrew-woosnam/impact-framework-vscode/releases, lmk if that doesn't work for whatever reason!

**Comment by jawache** [comment]
This is wonderful @andrew-woosnam how exciting :) 

I definitely think this is useful, esp. as IF manifest files become more complex over time.

I imagine to analyze manifest files we'd probably use different visualsation tools, are you thinking here more tools to help with development, debugging and diagnosis?

Some future features (we've been discussing partially with @heaversm in another thread) that might be interesting are:
 
- Perhaps outputting an legend/key node (calling it explain right now) which gives you a description of what each parameter actually is, the units, the source (plugin) to aid with debugging.

- Plugin metadata which gives you information about what inputs/outputs they expect, also perhaps the configuration options/choices?

**Reply by andrew-woosnam** [reply]
> _are you thinking here more tools to help with development, debugging and diagnosis?_

Yeah generally just trying to aid anyone wanting to work with IF, especially if they're computing stuff from manifest files on the command line (keep it all in the IDE!)

> _to analyze manifest files we'd probably use different visualsation tools_

Sounds cool! Has there been any talk of specifics around this?

> _a description of what each parameter actually is, the units, the source (plugin) [...] information about what inputs/outputs they expect, also perhaps the configuration options/choices_

100% on board with these (@josh-swerdlow and I have been chatting about how best to make this a reality). We've been trying to keep up with the new discussions on these topics and will comment on those threads / keep them top of mind for how we build this vsc extension
